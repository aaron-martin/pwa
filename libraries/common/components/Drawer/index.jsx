/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import styles from './style';

/**
 * Drawer component.
 */
class Drawer extends Component {
  /**
   * The component prop types.
   * @type {Object}
   */
  static propTypes = {
    alwaysActive: PropTypes.bool,
    animation: PropTypes.shape({
      duration: PropTypes.number,
      in: PropTypes.string,
      out: PropTypes.string,
    }),
    children: PropTypes.node,
    className: PropTypes.string,
    isOpen: PropTypes.bool,
    onClose: PropTypes.func,
    onDidClose: PropTypes.func,
    onOpen: PropTypes.func,
  };

  /**
   * The component default props.
   * @type {Object}
   */
  static defaultProps = {
    alwaysActive: false,
    className: '',
    children: null,
    isOpen: false,
    onOpen: () => {},
    onClose: () => {},
    onDidClose: () => {},
    animation: {
      duration: null,
      in: '',
      out: '',
    },
  };

  /**
   * Initializes the Drawer component.
   * @param {Object} props The components props.
   */
  constructor(props) {
    super(props);

    this.state = {
      active: props.isOpen,
    };
  }

  /**
   * Update state when isOpen changes.
   * @param {Object} nextProps The next component props.
   */
  componentWillReceiveProps(nextProps) {
    if (this.props.isOpen !== nextProps.isOpen) {
      if (nextProps.isOpen) {
        if (typeof nextProps.onOpen === 'function') {
          nextProps.onOpen();
        }
        this.setState({ active: true });
      } else if (typeof nextProps.onClose === 'function') {
        nextProps.onClose();
      }
    }
  }

  /**
   * Syncs the internal active state when an animation ends.
   */
  handleAnimationEnd = () => {
    this.setState({ active: this.props.isOpen });
    if (!this.props.isOpen) {
      this.props.onDidClose();
    }
  };

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    const {
      alwaysActive,
      className,
      children,
      isOpen,
      animation,
    } = this.props;
    const { active } = this.state;

    const animationIn = animation.in || styles.animation.in;
    const animationOut = animation.out || styles.animation.out;

    const combinedClassName = classNames(
      className,
      styles.container,
      { [animationIn]: isOpen },
      { [animationOut]: !isOpen }
    );

    const style = {};
    if (typeof animation.duration === 'number') {
      style.animationDuration = `${animation.duration}ms`;
    }

    return (active || alwaysActive) ? (
      <div
        className={combinedClassName}
        style={style}
        onAnimationEnd={this.handleAnimationEnd}
      >
        {children}
      </div>
    ) : null;
  }
}

export default Drawer;
