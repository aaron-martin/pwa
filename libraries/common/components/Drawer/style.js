/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { css } from 'glamor';

const duration = 150;
const easing = 'ease';

const slideInBaseDrawer = css.keyframes({
  '0%': { transform: 'translateY(100%)' },
  '100%': { transform: 'translateY(0)' },
});

const slideOutBaseDrawer = css.keyframes({
  '0%': { transform: 'translateY(0)' },
  '100%': { transform: 'translateY(100%)' },
});

const container = css({
  position: 'fixed',
  bottom: 0,
  left: 0,
  zIndex: 5,
}).toString();

const animation = {
  in: css({
    animation: `${slideInBaseDrawer} ${duration}ms 1 both ${easing}`,
  }).toString(),
  out: css({
    animation: `${slideOutBaseDrawer} ${duration}ms 1 both ${easing}`,
  }).toString(),
};

export default {
  container,
  animation,
};
