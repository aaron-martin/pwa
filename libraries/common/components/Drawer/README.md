# Drawer
---
This component transitions based on it's `isOpen` property. It is intended to be used to provide an additional UI layer to a page.

## Getting Started

```markup
import { Drawer } from 'Library/components';

<Drawer isOpen={true}>
  <SomeContent />
</Drawer>
```

## Props

### className

_Default_: `''`
_Type_: `string`

Additional classes that can be appended to the component's `className` attribute.

###### Usage:
```
<Drawer className="class1 class2">
  ...
</Drawer>
```

### children

_Default_: `null`
_Type_: `node`

Children that will be rendered.

###### Usage:

```
<Drawer>
  <ChildComponent />
</Drawer>
```

### isOpen

_Default_: `false`
_Type_: `boolean`

Controls whether or not the drawer is open.

###### Usage:

```
<Drawer isOpen={true}>
  ...
</Drawer>
```

### onOpen

_Default_: `() => {}`
_Type_: `Function`

A callback function that is fired after the Drawer has been opened.

###### Usage:

```
const myFunc = () => {
  ...
};

<Drawer onOpen={myFunc}>
  ...
</Drawer>
```

### onClose

_Default_: `() => {}`
_Type_: `Function`

A callback function that is fired after the Drawer has been closed.

###### Usage:

```
const myFunc = () => {
  ...
};

<Drawer onClose={myFunc}>
  ...
</Drawer>
```

### animation

_Default_:

```
{
  duration: 300,
  easing: 'ease-out',
  from: null,
  set: null,
  to: null,
}
```
_Type_: `Object`

An object that describes the transition of the Drawer.

###### Usage:

```
const drawerAnimation = {
  from: {
    y: '100%'
  },
  to: {
    y: '0%'
  },
};

<Drawer animation={drawerAnimation}>
  ...
</Drawer>
```
