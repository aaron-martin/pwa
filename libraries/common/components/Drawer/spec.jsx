/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount, shallow } from 'enzyme';
import Drawer from './index';

describe('<Drawer />', () => {
  let mockOpen;

  beforeEach(() => {
    mockOpen = jest.fn();
  });

  it('should render', () => {
    const wrapper = shallow(<Drawer />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should execute callback when drawer is opened', () => {
    const wrapper = mount(<Drawer onOpen={mockOpen} />);
    wrapper.setProps({
      isOpen: true,
    });

    expect(mockOpen).toBeCalled();
  });

  it('should execute callback when drawer is closed', () => {
    const wrapper = mount(<Drawer isOpen onClose={mockOpen} />);
    wrapper.setProps({
      isOpen: false,
    });

    expect(mockOpen).toBeCalled();
  });

  it('should add custom classes', () => {
    const wrapper = mount(<Drawer className="custom-class-name" isOpen />);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.hasClass('custom-class-name')).toEqual(true);
  });

  it('should execute callback when drawer close animation did end', () => {
    const wrapper = mount(<Drawer className="custom-class-name" isOpen onDidClose={mockOpen} />);
    expect(wrapper).toMatchSnapshot();
    wrapper.setProps({
      isOpen: false,
    });
    wrapper.simulate('animationEnd');
    expect(mockOpen).toBeCalled();
  });
});
