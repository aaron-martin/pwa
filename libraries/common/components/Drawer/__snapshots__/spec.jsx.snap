// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> should add custom classes 1`] = `
<Drawer
  alwaysActive={false}
  animation={
    Object {
      "duration": null,
      "in": "",
      "out": "",
    }
  }
  className="custom-class-name"
  isOpen={true}
  onClose={[Function]}
  onOpen={[Function]}
>
  <div
    className="custom-class-name css-1rg7udn css-1dwe0qx"
    onAnimationEnd={[Function]}
    style={Object {}}
  />
</Drawer>
`;

exports[`<Drawer /> should render 1`] = `null`;
